File: problem-1-binaryTree-levelOrder-Traversal.py
"""
Problem:
https://leetcode.com/problems/binary-tree-level-order-traversal/

Approach: BFS Solution

1. Maintain a queue
2. push nodes at each level to the queue
3. pop nodes from each level from the queue and add to a list and append the list to a global list

"""

# Time Complexity : O(n) where n is the no of nodes in the tree
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : yes	
# Any problem you faced while coding this : No


from collections import deque
class Solution:
    que = None
    res = None
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if root==None:
            return self.res
        self.res = []
        self.que = deque()
        self.que.append(root)
        while(self.que):
            qsize = len(self.que)
            count = 0
            temp = []
            while(count<qsize):
                popped = self.que.popleft()
                temp.append(popped.val)
                count+=1
                if (popped.left!=None):
                    self.que.append(popped.left) 
                if (popped.right!=None):
                    self.que.append(popped.right)
            self.res.append(temp)
        return self.res


"""
Problem:
https://leetcode.com/problems/binary-tree-level-order-traversal/

Approach: DFS Solution

1. For each node in the tree, store the level information of the node in the recursive stack
2. Maintain a global array, if we hit a node at a perticular level in the recursion and the level is equal to the size of the list then we make a new sub list and add the node to that sub list and append the sub list to global list
3. if the length of the global list is greater that than current level, take the sub list at current level and append the root.val to that sublist

"""

# Time Complexity : O(n) where n is the no of nodes in the tree
# Space Complexity : O(H) where H is the  height of the tree
# Did this code successfully run on Leetcode : yes	
# Any problem you faced while coding this : No

class Solution:
    res = None
    
    def dfs(self,root,level):
        if root==None:
            return
        
        if level==len(self.res):
            self.res.append([])
        
        self.res[level].append(root.val)
        self.dfs(root.left,level+1)
        self.dfs(root.right,level+1)
        
        
    
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        self.res = []
        if root == None:
            return self.res
        self.dfs(root,0)
        return self.res

File: Problem1_BinaryTreeLevelOrderTraversal.cpp

// Time Complexity : O(n) where n is number of elements
// Space Complexity : O(1) we are not using extra space
// Did this code successfully run on Leetcode :yes
// Did this code successfully run on Home system :No i got errors
//          Problem1_BinaryTreeLevelOrderTraversal.cpp:120:30: error: use of undeclared identifier 'levelOrder'
//                  vector<vector<int>> res =levelOrder(root); 


//          Problem1_BinaryTreeLevelOrderTraversal.cpp:122:14: error: invalid operands to binary expression ('std::__1::ostream'
    //   (aka 'basic_ostream<char>') and 'std::__1::__vector_base<std::__1::vector<int, std::__1::allocator<int> >,
    //   std::__1::allocator<std::__1::vector<int, std::__1::allocator<int> > > >::value_type' (aka 'std::__1::vector<int,
    //   std::__1::allocator<int> >'))
    //     cout << res[i] << " "; 

// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach


#include<vector>
#include<bits/stdc++.h>
#include<stdlib.h>
#include<cmath>
#include<time.h>
#include<iostream>

using namespace std;

 //* Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        
        vector<vector<int>> result;
        if(root==NULL) return result;
        
        queue<TreeNode*> q;
        q.push(root);
        
        while(!q.empty()){
            int count = q.size();
            vector<int> temp;
            for(int i=0;i<count;i++){
                TreeNode* curr= q.front(); 
                q.pop();
                temp.push_back(curr->val);
                if(curr->left != NULL) q.push(curr->left);
                if(curr->right != NULL) q.push(curr->right);
            }
            result.push_back(temp);
        }
        
        return result;
        
    }
};


//Below is DFS solution

// class Solution {
//     vector<vector<int>> result;
// public:
//     vector<vector<int>> levelOrder(TreeNode* root) {
        
        
//         if(root==NULL) return result;
        
//         // queue<TreeNode*> q;
//         // q.push(root);
//         // while(!q.empty()){
//         //     int size = q.size();
//         //     vector<int> temp;
//         //     for(int i=0;i<size;i++){
//         //         TreeNode* curr= q.front(); 
//         //         temp.push_back(curr->val);
//         //         if(curr->left != NULL) q.push(curr->left);
//         //         if(curr->right != NULL) q.push(curr->left);
//         //     }
//         //     result.push_back(temp);
//         // }
//         dfs(root,0);
//         return result;
        
//     }
// private:
//     void dfs(TreeNode* root, int level){
//         //base
//         if(root==NULL) return;
        
        
//         //logic
//         if(level==result.size()){
//             result.push_back(vector<int>());
//         }
//         result.at(level).push_back(root->val);
//         dfs(root->left,level+1);
//         //st.pop()
//         dfs(root->right,level+1);
//     }
// };

int main() 
{ 
    /* Let us construct a Binary Tree 
        4 
       / \ 
      9   2 
     / \   \ 
    3   5   7 */
  
    TreeNode* root = NULL; 
    root = new TreeNode(4); 
    root->left = new TreeNode(9); 
    root->right = new TreeNode(2); 
    root->left->left = new TreeNode(3); 
    root->left->right = new TreeNode(5); 
    root->right->right = new TreeNode(7); 
      
    vector<vector<int>> res =levelOrder(root); 
    for (int i = 0; i < res.size(); i++) 
        cout << res[i] << " "; 
          
    return 0; 
} 